src/Orleans/Threading/CancellationSourcesExtension.cs

```diff
@@ -0,0 +1,73 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Orleans.Concurrency;
+
+namespace Orleans.Threading
+{
+    internal class CancellationSourcesExtension : ICancellationSourcesExtension
+    {
+        private static readonly CancellationTokenSource CancelledTokenSource = new CancellationTokenSource(1);
+
+        private readonly ConcurrentDictionary<Guid, CancellationTokenSource> _cancellationTokenSources;
```

Comment:
I don't think you need ConcurrentDictionary here. You are creating one CancellationSourcesExtension  per activation, and activation is single threaded, so regular dictionary is OK. We do need to make sure we only access it from the activation context.